src/sign/keyset.rs

```diff
@@ -0,0 +1,1906 @@
+        let mut first = true;
+        if self.old {
+            write!(f, "Old")?;
+            first = false;
+        }
+        if self.signer {
+            write!(f, "{}Signer", if first { "" } else { ", " })?;
+            first = false;
+        }
+        if self.present {
+            write!(f, "{}Present", if first { "" } else { ", " })?;
+            first = false;
+        }
+        if self.at_parent {
+            write!(f, "{}At Parent", if first { "" } else { ", " })?;
+        }
+        match (self.old, self.signer, self.present) {
+            (false, false, false) => write!(f, "(Future)")?,
+            (false, false, true) => write!(f, " (Incoming)")?,
+            (false, true, true) => write!(f, " (Active)")?,
+            (true, true, true) => write!(f, " (Leaving)")?,
+            (true, false, true) => write!(f, " (Retired)")?,
+            (true, false, false) => write!(f, " (Old)")?,
+            (_, _, _) => (),
+        }
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+/// This type contains the various timestamps in the life of a key.
+///
+/// All timestamps are optional. The following timestamps are supported:
+/// * when the key was created,
+/// * when the key was first published in the DNSKEY RRset,
+/// * when the DNSKEY RRset with the key was first visible,
+/// * when the DS record for the key was first visible,
+/// * when RRSIG records signed by the key were first visible,
+/// * when the key was withdrawn from the DNSKEY RRset.
+pub struct KeyTimestamps {
+    creation: Option<UnixTime>,
+    published: Option<UnixTime>,
+    visible: Option<UnixTime>,
+    ds_visible: Option<UnixTime>,
+    rrsig_visible: Option<UnixTime>,
+    withdrawn: Option<UnixTime>,
+}
+
+impl KeyTimestamps {
+    /// Return the creation time of a key.
+    pub fn creation(&self) -> Option<UnixTime> {
+        self.creation.clone()
+    }
+
+    /// Return the time when a key was first published in the DNSKEY RRset.
+    pub fn published(&self) -> Option<UnixTime> {
+        self.published.clone()
+    }
+
+    /// Return the time when DNSKEY RRset with a key was first visible.
+    pub fn visible(&self) -> Option<UnixTime> {
+        self.visible.clone()
+    }
+
+    /// Return the time when a DS record for a key was first visible.
+    pub fn ds_visible(&self) -> Option<UnixTime> {
+        self.ds_visible.clone()
+    }
+
+    /// Return the time when an RRSIG record signed by the key was first
+    /// visible.
+    pub fn rrsig_visible(&self) -> Option<UnixTime> {
+        self.rrsig_visible.clone()
+    }
+
+    /// Return the time when the key was removed from the DNSKEY RRset.
+    pub fn withdrawn(&self) -> Option<UnixTime> {
+        self.withdrawn.clone()
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+/// A type that contains Unix time.
+///
+/// Unix time is the number of seconds since midnight January first
+/// 1970 GMT.
+pub struct UnixTime(Duration);
```

Comment:
The reason is serde. I want to make sure that the time goes out as seconds since unix epoch. I don't think SystemTime guarantees that. After discussion with @partim we concluded that a type that is clearly defined as Unix time could be useful in more places.

Reply:
That's understandable.  But I had a peek at serde: the [`impl Serialize for SystemTime`](https://docs.rs/serde/latest/src/serde/ser/impls.rs.html#754-767) shows that it serializes as seconds and nanoseconds since the Unix Epoch.  I'm sure they can't change this now, for backward compatibility, so you can rely on it.

Reply:
We can change that later if it makes sense. I don't like relying on undocumented features. It is also important that any time related type is easy to mock.